File: leetcode/length-of-last-word.py
# 58. Length of Last Word
# 🟢 Easy
#
# https://leetcode.com/problems/length-of-last-word/
#
# Tags: String

import timeit


# Use the built-in str.split function to get the last word.
#
# Time complexity: O(n) - The function iterates over all characters in
# the input.
# Space complexity: O(n) - The function creates an array with all the
# words in the input.
#
# Runtime 33 ms Beats 73.47%
# Memory 13.8 MB Beats 76.78%
class BuiltIn:
    def lengthOfLastWord(self, s: str) -> int:
        return len(s.split()[-1])


# Start iterating over the characters in the input string starting from
# the back, first ignore all whitespace, then count characters, as soon
# as we see whitespace again, stop iterating and return the length of
# the word.
#
# Time complexity: O(n) - We may iterate over all the input.
# Space complexity: O(1) - We only use pointers.
#
# Runtime 33 ms Beats 85.69%
# Memory 13.8 MB Beats 76.78%
class Solution:
    def lengthOfLastWord(self, s: str) -> int:
        res = 0
        for i in reversed(range(len(s))):
            # Skip any whitespace while we don't have a word.
            if s[i] == " ":
                # If we already computed a word's length, return it.
                if res:
                    return res
                # If this is whitespace at the end, ignore it.
                continue
            res += 1
        return res


def test():
    executors = [
        BuiltIn,
        Solution,
    ]
    tests = [
        ["l     ", 1],
        ["     e", 1],
        ["Hello World", 5],
        ["luffy is still joyboy", 6],
        ["   fly me   to   the moon  ", 4],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.lengthOfLastWord(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/length-of-last-word.rs
// 58. Length of Last Word
// 🟢 Easy
//
// https://leetcode.com/problems/length-of-last-word/
//
// Tags: String

struct Solution;
impl Solution {
    // Iterate over the characters in the input string in reverse order, ignore any whitespace
    // found before non-whitespace chars, then count the non-whitespace chars until more whitespace
    // is found or we get to the beginning of the string.
    //
    // Time complexity: O(n) - We may iterate over all the input.
    // Space complexity: O(1) - We only use pointers.
    //
    // Runtime 1 ms Beats 71.17%
    // Memory 2.1 MB Beats 34.67%
    #[allow(dead_code)]
    pub fn length_of_last_word_for(s: String) -> i32 {
        // Count of characters in the last word.
        let mut count = 0;
        for char in s.chars().rev() {
            if char == ' ' {
                if count == 0 {
                    continue;
                }
                return count;
            }
            count += 1;
        }
        // If we exit the loop is because the last word starts at index 0
        count
    }

    /// Same logic using iterators.
    ///
    /// Time complexity: O(n) - We iterate over the last word and the following whitespace.
    /// Space complexity: O(1) - Constant extra space used.
    ///
    /// Runtime 0 ms Beats 100%
    /// Memory 2.14 MB Beats 25%
    pub fn length_of_last_word(s: String) -> i32 {
        s.trim_end()
            .chars()
            .rev()
            .take_while(|c| c.is_alphabetic())
            .count() as i32
    }
}

// Tests.
fn main() {
    let tests = [
        ("", 0),
        ("   ", 0),
        ("l  ", 1),
        ("Hello World", 5),
        ("luffy is still joyboy", 6),
        ("   fly me     to     the     moon    ", 4),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::length_of_last_word(t.0.to_string());
        if res == t.1 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.1, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
